### ISSUE TITLE
Feature: Make `/bake` time-to-live configurable & by default 0

### ISSUE THREAD

**Issue by jpmcb** [open-sauced/pizza#16]
### Type of feature

🍕 Feature

### Current behavior

Currently, when making a request to the `/bake` route, the connection stays alive for as long as it will takes to process a repo. This may take abit (maybe 30 seconds) depending on the size of the repo.

### Suggested solution

The time to live (TTL) should be made configurable so that the pizza server can hangup connections in the case that users do not want to wait.

This can also give us the opportunity to flip the default experience where the server can automatically hangup connections and return 200 upon receiving a valid request. I.e., the server, by default, shouldn't keep connections alive unless a TTL has been set by the user.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Contributing Docs

- [X] I agree to follow this project's Contribution Docs

**Comment by AvineshTripathi** [comment]
I would love to take this!

**Comment by k1nho** [comment]
I think a good option could be to define a struct config, then make **PizzaOvenServer** wrap the config. I think config struct in the long run has many advantages including configuration for max open db connections, max idle db connections, and db connections max idle time.

```go
type config struct {
	TTL time.Duration
}

type PizzaOvenServer struct {
	config    config
	Logger    *zap.SugaredLogger
	PizzaOven *database.PizzaOvenDbHandler
}
```

Then, parse TTL as flag (or in .env), like so when calling **NewPizzaOvenServer**
```go
func NewPizzaOvenServer(dbHandler *database.PizzaOvenDbHandler) *PizzaOvenServer {
        // configure Time to Live 
	var cfg config
	flag.DurationVar(&cfg.TTL, "ttl", 0*time.Second, "Time to Live(ex: 30s)")
	flag.Parse()

	logger, err := zap.NewProduction()
	if err != nil {
		log.Fatalf("Could not initiate zap logger: %v", err)
	}
	sugarLogger := logger.Sugar()
	sugarLogger.Infof("initiated zap logger with level: %d", sugarLogger.Level())

	return &PizzaOvenServer{
		Logger:    sugarLogger,
		PizzaOven: dbHandler,
		config:    cfg,
	}
}
```
After, launching the server, if no parameter is provided then it just defaults to the one defined in flag.

```bash
go run main.go -ttl=30s
```

Then, we can make use of config.TTL to provide a context for DB queries and cancel, if it exceeds config.TTL. Though, I'm not sure how TTL would work for **processRepository** since it does 3 DB calls, is TTL granular to each DB call or should be an aggregation?

**Comment by k1nho** [comment]
> Since some repos with many thousands of commits may take awhile to process, we need some mechanism to hangup the connection without disrupting the batch process of commits.

Oh ok, yeah this makes sense. If I understood, this should not necessarily stop the DB queries in process, but just make the server hang up the connection (after some time) for the client to continue with another call to bake.

> Hmm, but maybe a better approach is for the server to validate the request and just immediately return 200 without trying to keep the connection open at all. This service is essentially a batch processor of git commits so maybe it doesn't make sense for clients to wait for those batch processes to finish.

This sounds great, but then it would need another way to communicate to the client if the batch process was a success, or if it failed (this would be the case either way for very low ttl trying to process Linux repo or something of that magnitude).

**Comment by AvineshTripathi** [comment]
looking at the usecase as I suggested about the tickers could be good. So your ttl could be a flag `default=0` and based on this value we can start a ticker. 

```

func (p PizzaOvenServer) handleRequest(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		p.Logger.Errorf("Received request with invalid method: %v", r.Body)
		http.Error(w, "Invalid request method, expected post", http.StatusMethodNotAllowed)
		return
	}

	var data reqData
	err := json.NewDecoder(r.Body).Decode(&data)
	if err != nil {
		p.Logger.Errorf("Could not decode request json body: %v with error: %v", r.Body, err)
		http.Error(w, "Could not decode request body", http.StatusBadRequest)
		return
	}

	err = p.processRepository(data.URL)
	if err != nil {
		p.Logger.Errorf("Could not process repository input: %v with error: %v", r.Body, err)
		http.Error(w, "Could not process input", http.StatusInternalServerError)
		return
	}
        if ttl != 0 {
                startticker(ttl) // eqvivalent func to this
        }
        return 
}

```

Just a rough idea. Suggestions are welcome

**Comment by jpmcb** [comment]
Right. If the client has a very long TTL, it should return a server error if there happens to be a problem while git commits are processing (which is essentially the behavior now).

A status endpoint would be really tricky to get right since this service can be scaled horizontally behind a load balancer where any single instance may be processing a request. One solution would be to deploy a balancer with sticky sessions but that's out of scope of this service and gets too into the weeds of how this software may be run.

Maybe a better behavior is for a kubectl like experience with the --wait flag / command:
- by default, the server validates the request and returns the status. It processes commits in the background.
- if a --wait flag / request json field is set, the server keeps the connection alive until processing is completed or fails.

**Comment by k1nho** [comment]
Wait flag looks like a great idea. Current behavior for **/bake** does not have **200 OK** on success, is this something that would be expected or is **202** the expected? Then, write success when no err. Something like this maybe?

```go
func (p PizzaOvenServer) handleRequest(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		p.Logger.Errorf("Received request with invalid method: %v", r.Body)
		http.Error(w, "Invalid request method, expected post", http.StatusMethodNotAllowed)
		return
	}

	var data reqData
	err := json.NewDecoder(r.Body).Decode(&data)
	if err != nil {
		p.Logger.Errorf("Could not decode request json body: %v with error: %v", r.Body, err)
		http.Error(w, "Could not decode request body", http.StatusBadRequest)
		return
	}

	w.WriteHeader(http.StatusAccepted)
	if p.config.Wait {
		err = p.processRepository(data.URL)
		if err != nil {
			p.Logger.Errorf("Could not process repository input: %v with error: %v", r.Body, err)
			http.Error(w, "Could not process input", http.StatusInternalServerError)
			return
		}
	} else {
		go func() {
			err = p.processRepository(data.URL)
			if err != nil {
				p.Logger.Errorf("Could not process repository input: %v with error: %v", r.Body, err)
				http.Error(w, "Could not process input", http.StatusInternalServerError)
			}
		}()
	}
}
```

**Comment by AvineshTripathi** [comment]
@k1nho approach LGTM, however inplace of directly sending a 200, we should check atleast if the URL is valid what do you think? 

Also might be too early to say that or maybe I am understanding in wrong way but I would like to point this that going with go routine knowing to the fact that we might use go routine inside this anonymous go routine to speed up things. This can lead to memory problems later.

Like to know your thoughts....

**Comment by k1nho** [comment]
>  we should check atleast if the URL is valid what do you think?

Yes, I was thinking maybe a validator package can be created that can be reused across the different queries ?

> Also might be too early to say that or maybe I am understanding in wrong way but I would like to point this that going with go routine knowing to the fact that we might use go routine inside this anonymous go routine to speed up things. This can lead to memory problems later.

You are right, as I understand the queries happen sequentially at the moment, but they might be thrown in a go routine in the future. Another problem might be just throwing a bunch of burst request for very large repos like Linux, that will fill up memory very fast, a limit might be necessary related to #17 perhaps?

**Comment by jpmcb** [comment]
> Yes, I was thinking maybe a validator package can be created that can be reused across the different queries ?

👍🏼 

> Also might be too early to say that or maybe I am understanding in wrong way but I would like to point this that going with go routine knowing to the fact that we might use go routine inside this anonymous go routine to speed up things. This can lead to memory problems later.

The `http` servermux already services requests via concurrent go routines. And you can see this at work by making a request like this to a locally running server:
```
❯ curl -d '{"url":"https://github.com/open-sauced/insights"}' \
    -H "Content-Type: application/json" \
    -X POST http://localhost:8080/bake
```
this will take a minute to process and while that's happening, you can hit the `/ping` route many many times and it is serviced just like any other route:
```
❯ curl localhost:8080/ping
```

I've been keeping concurrency in mind for this server and #17 addresses that as well.

---

If one of you want to pick this up, I'd be happy to review. Great discussion!

**Comment by k1nho** [comment]
I can give this one a try

**Comment by jpmcb** [comment]
Closed by #21
